File: Algorithms/Graphs/BoruvkaMST.java

import java.util.Arrays;

public class BoruvkaMST {

    private static class Edge {

        int dist, fromVertex, toVertex;

        public Edge(int from, int to, int weight) {

            fromVertex = from;
            toVertex = to;
            dist = weight;

        }

    }

    private static Edge[] getEdges(int[][] graph, int numEdges) {

        Edge[] edges = new Edge[numEdges];

        int k = 0;

        for (int i = 0; i < graph.length; i++) {

            for (int j = i + 1; j < graph.length; j++) {

                if (graph[i][j] > 0) {
                    edges[k++] = new Edge(i, j, graph[i][j]);
                }

            }

        }

        return edges;

    }

    private static int findGroup(int[] parents, int i) {

        while (i != parents[i]) {

            i = parents[i];

        }

        return i;

    }

    public static void printMST(int[][] graph, int numEdges) {

        int n = graph.length, resultIndex = 0;

        Edge[] edges = getEdges(graph, numEdges);

        Edge[] result = new Edge[n];

        int[] closestIndex = new int[n];

        int[] parents = new int[n];

        for (int i = 0; i < n; i++) {
            parents[i] = i;
        }

        while (resultIndex < n - 1) {

            Arrays.fill(closestIndex, -1);

            for (int i = 0; i < edges.length; i++) {

                int a = findGroup(parents, edges[i].fromVertex);
                int b = findGroup(parents, edges[i].toVertex);

                if (a != b) {

                    if (closestIndex[a] == -1 || edges[closestIndex[a]].dist > edges[i].dist) {

                        closestIndex[a] = i;

                    }

                    if (closestIndex[b] == -1 || edges[closestIndex[b]].dist > edges[i].dist) {

                        closestIndex[b] = i;

                    }

                }

            }

            for (int i = 0; i < n; i++) {

                if (closestIndex[i] != -1) {

                    int a = findGroup(parents, edges[closestIndex[i]].fromVertex);
                    int b = findGroup(parents, edges[closestIndex[i]].toVertex);

                    if (a != b) {

                        result[resultIndex++] = edges[closestIndex[i]];

                        parents[a] = b;

                    }

                }

            }

        }

        for (int i = 0; i < resultIndex; i++) {

            System.out.printf("%d -> %d%n", result[i].fromVertex, result[i].toVertex);

        }

    }

    public static void main(String[] args) {

        int[][] graph = new int[][]{
            {0, 3, 6, 5},
            {3, 0, 0, 9},
            {6, 0, 0, 4},
            {5, 9, 4, 0}
        };

        printMST(graph, 5);

    }

}

File: Algorithms/Graphs/BoruvkaMST.js
"use strict";

class Edge {

    constructor(from, to, weight) {

        this.fromVertex = from;
        this.toVertex = to;
        this.dist = weight;

    }

}

let getEdges = (graph) => {

    let edges = [];

    for (let i = 0; i < graph.length; i++) {

        for (let j = i + 1; j < graph.length; j++) {

            if (graph[i][j] > 0) {
                edges.push(new Edge(i, j, graph[i][j]));
            }

        }

    }

    return edges;

}

let _findGroup = (parents, node) => {

    while (node != parents[node]) {
        node = parents[node];
    }

    return node;

}

let printMST = (graph) => {

    let edges = getEdges(graph);
    let n = graph.length;
    let result = [],
        parents = [...Array(n).keys()];

    while (result.length < n - 1) {

        let closestIndex = (new Array(n)).fill(-1);

        for (let i = 0; i < edges.length; i++) {



            let a = _findGroup(parents, edges[i].fromVertex),
                b = _findGroup(parents, edges[i].toVertex);

            if (a != b) {

                if (closestIndex[a] == -1 || edges[closestIndex[a]].dist > edges[i].dist) {

                    closestIndex[a] = i;

                }

                if (closestIndex[b] == -1 || edges[closestIndex[b]].dist > edges[i].dist) {

                    closestIndex[b] = i;

                }

            }

        }

        for (let index of closestIndex) {

            if (index != -1) {

                let a = _findGroup(parents, edges[index].fromVertex),
                    b = _findGroup(parents, edges[index].toVertex);

                if (a != b) {

                    result.push(edges[index]);

                    parents[a] = b;

                }

            }

        }

    }

    for (let edge of result) {

        console.log(`${edge.fromVertex} -> ${edge.toVertex}`);

    }

}

let graph = [
    [0, 3, 6, 5],
    [3, 0, 0, 9],
    [6, 0, 0, 4],
    [5, 9, 4, 0]
];

printMST(graph);
